File: jlox/src/main/java/fr/ocus/lox/jlox/Environment.java
package fr.ocus.lox.jlox;

import java.util.HashMap;
import java.util.Map;

/**
 * @author Matthieu Honel <ocus51@gmail.com>
 * @since 2017-08-03
 */
public class Environment {
    final Environment enclosing;
    private final Map<String, Object> values = new HashMap<>();

    Environment() {
        this(null);
    }

    Environment(Environment enclosing) {
        this.enclosing = enclosing;
    }

    Map<String, Object> getValues() {
        return values;
    }

    Map<String, Object> getAllValues() {
        Map<String, Object> allValues = new HashMap<>();
        Environment environment = this;
        do {
            allValues.putAll(environment.values);
            environment = environment.enclosing;
        } while (environment != null);

        return allValues;
    }

    void define(String name, Object value) {
        values.put(name, value);
    }

    void assign(Token name, Object value) {
        if (values.containsKey(name.lexeme)) {
            values.put(name.lexeme, value);
            return;
        }

        if (enclosing != null) {
            enclosing.assign(name, value);
            return;
        }

        throw new RuntimeError(name, "Undefined variable '" + name.lexeme + "'.");
    }

    Object get(Token name) {
        if (values.containsKey(name.lexeme)) {
            return values.get(name.lexeme);
        }

        if (enclosing != null) return enclosing.get(name);

        throw new RuntimeError(name, "Undefined variable '" + name.lexeme + "'.");
    }

    void assignAt(int distance, Token name, Object value) {
        Environment environment = this;
        for (int i = 0; i < distance; i++) {
            environment = environment.enclosing;
        }

        environment.values.put(name.lexeme, value);
    }

    Object getAt(int distance, String name) {
        Environment environment = this;
        for (int i = 0; i < distance; i++) {
            environment = environment.enclosing;
        }

        return environment.values.get(name);
    }

    @Override
    public String toString() {
        String result = values.toString();
        if (enclosing != null) {
            result += " -> " + enclosing.toString();
        }

        return result;
    }
}

File: pylox/pylox/environment.py
from .error import PyLoxRuntimeError
from .interface import EnvironmentInterface
from .token import Token


class Environment(EnvironmentInterface):
    def __init__(self, enclosing: 'Environment' = None):
        self.enclosing = enclosing
        self._values = {}

    def define(self, name: str, value):
        self._values[name] = value

    def assign(self, token: Token, value):
        if token.lexeme in self._values:
            self._values[token.lexeme] = value
            return

        if self.enclosing:
            self.enclosing.assign(token, value)
            return

        raise PyLoxRuntimeError(token=token, message="Undefined variable '" + token.lexeme + "'.")

    def get(self, token: Token):
        if token.lexeme in self._values:
            return self._values[token.lexeme]

        if self.enclosing:
            return self.enclosing.get(token=token)

        raise PyLoxRuntimeError(token=token, message="Undefined variable '" + token.lexeme + "'.")

    def assign_at(self, distance: int, token: Token, value):
        environment = self
        for i in range(0, distance):
            environment = environment.enclosing

        environment._values[token.lexeme] = value

    def get_at(self, distance: int, name: str):
        environment = self
        for i in range(0, distance):
            environment = environment.enclosing

        return environment._values[name]

    def __str__(self):
        result = str(self._values)
        if self.enclosing:
            result += ' -> ' + str(self.enclosing)
        return result
